### DISCUSSION TITLE
Unable to flash Arduino UNO R4 Minima

### DISCUSSION THREAD

**Discussion by jomoengineer** [zephyrproject-rtos/zephyr#64813] [General]
I am trying the new examples with the Arduino UNO R4 Minima but the flash step is hanging.
I've gone through the install of Zephyr on a Ubuntu 20.04 VM using the Python virtual env and can build the code just fine. However, when flashing, it just shows: 

```
 (.venv) zephyrdev@ubuntu:~/zephyrproject/zephyr$ west flash
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner pyocd
-- runners.pyocd: Flashing file: /home/zephyrdev/zephyrproject/zephyr/build/zephyr/zephyr.hex
Waiting for a debug probe to be connected...
```
This is how the R4 shows from dmesg:
```
[ 1395.489518] usb 2-2.1: new full-speed USB device number 8 using uhci_hcd
[ 1395.807057] usb 2-2.1: New USB device found, idVendor=2341, idProduct=0069, bcdDevice= 1.00
[ 1395.807075] usb 2-2.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 1395.807083] usb 2-2.1: Product: UNO R4 Minima
[ 1395.807089] usb 2-2.1: Manufacturer: Arduino
[ 1395.807094] usb 2-2.1: SerialNumber: 3502290C39313631047133324B572D76
[ 1395.820167] cdc_acm 2-2.1:1.0: ttyACM0: USB ACM device
```
Any suggestions on how to get past this?

**Comment by kartben** [comment]
I am assuming you installed the required CMSIS-Pack as per https://docs.zephyrproject.org/latest/boards/arm/arduino_uno_r4/doc/index.html#using-pyocd?

**Reply by jomoengineer** [reply]
Yes, I did.

**Comment by kartben** [comment]
@soburi any thoughts on this?

**Comment by jomoengineer** [comment]
Out of curiosity, am I supposed to have a particular debug probe connected to the Minima or do I just use the USB-C connection from the board like the Arduino IDE used?

**Comment by jomoengineer** [comment]
In either normal or DFU mode, pyocd is not finding a debug port for the R4:

```
(.venv) jomoros@jomoros-virtual-machine:~/zephyrproject/zephyr$ pyocd list
No available debug probes are connected
(.venv) jomoros@jomoros-virtual-machine:~/zephyrproject/zephyr$ pyocd list
No available debug probes are connected
```

However, I can see that the ports are showing in demsg. 

```
[  535.529055] usb 1-2.1: USB disconnect, device number 10
[  537.036681] usb 1-2.1: new full-speed USB device number 11 using uhci_hcd
[  537.342020] usb 1-2.1: New USB device found, idVendor=2341, idProduct=0369, bcdDevice= 1.00
[  537.342026] usb 1-2.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  537.342028] usb 1-2.1: Product: Santiago DFU
[  537.342030] usb 1-2.1: Manufacturer: Arduino
[  537.342031] usb 1-2.1: SerialNumber: 3502290C39313631047133324B572D76
[  814.156192] usb 1-2.1: USB disconnect, device number 11
[  815.785632] usb 1-2.1: new full-speed USB device number 12 using uhci_hcd
[  816.091557] usb 1-2.1: New USB device found, idVendor=2341, idProduct=0069, bcdDevice= 1.00
[  816.091563] usb 1-2.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  816.091566] usb 1-2.1: Product: UNO R4 Minima
[  816.091568] usb 1-2.1: Manufacturer: Arduino
[  816.091569] usb 1-2.1: SerialNumber: 3502290C39313631047133324B572D76
[  816.098561] cdc_acm 1-2.1:1.0: ttyACM0: USB ACM device

```

One thing I did find, was that I had to create the following USB rules for the Minima on Linux to even flash the board from the Arduino IDE as a regular user and not sudo. I found this after some searching.:
sudo view /etc/udev/rules.d/99-arduino-uno-r4.rules


```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="2341", ATTRS{idProduct}=="0069", GROUP="plugdev", MODE="0666"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="2341", ATTRS{idProduct}=="0369", GROUP="plugdev", MODE="0666"

```


It still does not work with west flash though.

**Comment by soburi** [comment]
First of all, I would like to check: Are you connecting the UNO R4 with an SWD cable? It's the small 2x5 connector on the top right.
Need to connect with debug adapter (CMSIS-DAP compatible adapter that can work with 5v). 
Currently, debugging or writing is not possible with a USB-C connection. UNO R4 uses a special dfu-util, so it is not yet supported.
I used a picoprobe (https://github.com/raspberrypi/picoprobe) and 5v-3.3v level-shifter in development.

**Comment by jomoengineer** [comment]
Yeah, that was one of my earlier questions, whether or not I needed a debugger for this.
Okay, that explains it.

I have a couple of Picos so I'll look into that.

Thanks.

**Reply by soburi** [reply]
As I was wrote, uno r4 io voltage is 5v, but pico is 3.3v. It's needs level shifter. Please  handle carefully.

**Comment by ClemensAtElektor** [comment]
On Windows 11 with the Arduino IDE 2 and the Minima board properly installed, you can use the command below over USB-C instead of using `west flash` with a programmer. Of course, make sure that the paths to `dfu-util` and `zephyr.bin` are correct. Put the Minima in DFU mode by pressing RESET twice. The LED should start to "breath".

`"%LOCALAPPDATA%\Arduino15\packages\arduino\tools\dfu-util\0.11.0-arduino5/dfu-util" --device 0x2341:0x0069,:0x0369 -D "build\zephyr\zephyr.bin" -a0 -Q`

Unfortunately, you cannot put this command in the `board.cmake` file for the Minima, as the script doesn't like the `dfu-util` arguments `-a0 -Q`. Maybe they can be mapped to other arguments?

To restore the Minima to "Arduino mode" (i.e. with bootloader), put the board in DFU mode and upload a sketch (any, e.g. Blink) with the IDE. The method described in the Zephyr Arduino Minima documentation doesn't work for me (using a J-Link Base pod). Note that trying to restore the bootloader with the command given above doesn't work either. There appears to be missing a programming step or something.
